internal/tpl/funcs.go

```diff
@@ -48,33 +48,30 @@ func sha1sum() func(...string) (string, error) {
 }
 
 // saltLen tries to parse the given string into a numeric salt length.
-// NOTE: This is on of the rare cases where I think named returns
-// are useful.
-func saltLen(s []string) (saltLen int) {
-	defer func() {
-		debug.Log("using saltLen %d", saltLen)
-	}()
-
+func saltLen(s []string) uint8 {
 	// default should be 32bit
-	saltLen = 32
+	sl := uint8(32)
+
+	defer func(sl uint8) {
+		debug.Log("using saltLen %d", sl)
+	}(sl)
 
 	if len(s) < 2 {
-		return
+		return sl
 	}
 
-	i, err := strconv.Atoi(s[0])
-	if err == nil && i > 0 {
-		saltLen = i
-	}
+	i, err := strconv.ParseUint(s[0], 10, 8)
```

Comment:
Is this really supposed to be `s[0]` and not `s[1]` ? Looking below at the usage with `ssha.Generate`, it seems `s[0]` is meant to be the password already...

Reply:
I'm very confused right now. It looks like it's been wrong all the time (i.e. the salt being the first parameter). So the hashes shouldn't work. But I'm quite sure I did use them for real. Which can't be true.

Let me try to add proper tests and then come back to this question.